### PR TITLE
Add file watch to support config reload on file change

### PR DESCRIPTION
**Pull request by vjsamuel**
**Description:**
Fixes: https://github.com/open-telemetry/opentelemetry-collector/issues/4397

This PR allows the main config.yml to be reloaded each time the config file changes. The entire pipeline gets reloaded.

This PR doesn't use an FS notify/inotify style watcher as we have seen that Kubernetes doesnt support such watches when a config is mounted as a config file. os.Stat given that it is cheap can be run every second to trigger a reload.

### CODE REVIEW

File: config/configmapprovider/file_watch.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	)
+
+	check := func() error {
+		currfi, err := os.Stat(filename)
+		modified := false
+		if err != nil {
+			if os.IsNotExist(err) && lastfi != nil {
+				return errNoOp
+			}
+			return err
```
Comment (tigrannajaryan) [inline config/configmapprovider/file_watch.go:40]: This appears to trigger reloading. Why do we reload if we can't stat the file?

File: config/configmapprovider/file_watch.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+			return nil
+		}
+
+		return errNoOp
+	}
+
+	// Perform an initial check.
+	err := check()
+	if err != nil && err != errNoOp {
+		return err
```
Comment (tigrannajaryan) [inline config/configmapprovider/file_watch.go:57]: Why do we give up watching if the initial check fails? I think we can keep checking and reload when a change is detected.

File: config/configmapprovider/file_watch.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				// If check returns a valid event, exit the loop. A new watch will be placed on the next Retrieve()
+				err := check()
+				if err == nil || err != errNoOp {
+					onChange(&ChangeEvent{
```
Comment (tigrannajaryan) [inline config/configmapprovider/file_watch.go:72]: This will reload the config immediately when a change is detected. That's an undesirable behavior since the file may be in the middle of being written and we may read partially written file. It is better to wait for some small amount of time (e.g. 1 second) after the last change to the file and only after that trigger reloading to increase the chance that the entire content of the file is written.

File: service/collector.go [hunk 4]
```diff
@@ -100,16 +89,10 @@ func New(set CollectorSettings) (*Collector, error) {
...
 
 	return &Collector{
-		set:          set,
-		stateChannel: make(chan State, Closed+1),
+		set: set,
 	}, nil
 }
 
-// GetStateChannel returns state channel of the collector server.
-func (col *Collector) GetStateChannel() chan State {
```
Comment (tigrannajaryan) [inline service/collector.go:109]: Why is this deleted? It is used in the tests.

File: service/collector.go [hunk 5]
```diff
@@ -137,21 +120,16 @@ func (col *Collector) runAndWaitForShutdownEvent(ctx context.Context) error {
...
 
 	col.shutdownChan = make(chan struct{})
-	col.stateChannel <- Running
 LOOP:
 	for {
 		select {
 		case err := <-col.cfgW.watcher:
 			col.logger.Warn("Config updated", zap.Error(err))
 
-			col.stateChannel <- Closing
```
Comment (tigrannajaryan) [inline service/collector.go:147]: I assume this was the problem that required deleting the code, right? It appears we will get blocked on writing after a config reloads.
We shouldn't delete the state channel and related code. It needs to be refactored to prevent blocking when not in testing mode (simplest is to just write only if there is room in the channel).

**Comment by tigrannajaryan** [comment]
> This PR doesn't use an FS notify/inotify style watcher as we have seen that Kubernetes doesnt support such watches when a config is mounted as a config file. os.Stat given that it is cheap can be run every second to trigger a reload.

Can we measure how much of CPU polling every second uses exactly?

File: service/collector.go (hunk 4 above)
Comment (vjsamuel) [reply service/collector.go:109]: Apologies for not mentioning this on the PR. This channel, given that it is used only in tests and has a size of 3, fills up after a couple of changes in the config and blocks the code flow when we try to add another state in. So i removed it to test and seek clarity on how to move forward.

File: config/configmapprovider/file_watch.go (hunk 3 above)
Comment (vjsamuel) [reply config/configmapprovider/file_watch.go:72]: ack.

File: service/collector.go (hunk 5 above)
Comment (vjsamuel) [reply service/collector.go:147]: ack.

File: config/configmapprovider/file_watch.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	"errors"
+	"os"
+	"time"
+)
+
+var errNoOp = errors.New("no change")
+
+// watchFile sets up a watch on a filename to detect changes and calls onChange() with a suitable ChangeEvent.
+// The watch is cancelled when the context is done.
+func watchFile(ctx context.Context, filename string, onChange func(*ChangeEvent)) error {
```
Comment (bogdandrutu) [inline config/configmapprovider/file_watch.go:28]: This does not need to be in a separate file. Consider to have it in "file.go"

File: service/collector.go (hunk 4 above)
Comment (bogdandrutu) [reply service/collector.go:109]: Please create a separate PR if this causes problems :)

File: config/configmapprovider/file_watch.go (hunk 2 above)
Comment (vjsamuel) [reply config/configmapprovider/file_watch.go]: i have removed this.

File: config/configmapprovider/file_watch.go (hunk 3 above)
Comment (vjsamuel) [reply config/configmapprovider/file_watch.go]: addressed by adding a sleep.

File: config/configmapprovider/file_watch.go (hunk 6 above)
Comment (vjsamuel) [reply config/configmapprovider/file_watch.go]: done

File: service/collector.go (hunk 4 above)
Comment (vjsamuel) [reply service/collector.go]: https://github.com/open-telemetry/opentelemetry-collector/pull/4460 -> done

**Comment by vjsamuel** [comment]
Raised https://github.com/open-telemetry/opentelemetry-collector/pull/4460 to fix the blocking channel

File: config/configmapprovider/file_watch.go (hunk 1 above)
Comment (vjsamuel) [reply config/configmapprovider/file_watch.go]: fixed this one. thanks for catching this. as a result, i changed the flow up slightly to ensure that if someone writes a faulty config, we preserve the last sane state as long as we dont shut down the process. Such a bug would bring down the collector across the entire kube cluster if there was a faulty config map update.

**Comment by vjsamuel** [comment]
@tigrannajaryan the benchmark that I have done was using the following code:
```
func BenchmarkOsStat(b *testing.B) {
	file, err := ioutil.TempFile("", "file_watcher_test")
	require.NoError(b, err)

	defer os.Remove(file.Name())
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		os.Stat(file.Name())
	}
}
```

and the result was:
```
goos: darwin
goarch: amd64
pkg: go.opentelemetry.io/collector/config/configmapprovider
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkOsStat
BenchmarkOsStat-12    	  197296	      6127 ns/op	     288 B/op	       2 allocs/op
PASS
```

we use this logic in some of our processing intensive code flows internally and we haven't run into issues so far.

**Comment by vjsamuel** [comment]
i have marked this PR as ready to review. it will require https://github.com/open-telemetry/opentelemetry-collector/pull/4460 to be reviewed and rebased with before this one can go in if approved.

File: config/configmapprovider/file_watch.go (hunk 3 above)
Comment (tigrannajaryan) [reply config/configmapprovider/file_watch.go]: I don't see where this is addressed, please point me to the code.

Comment (vjsamuel) [reply config/configmapprovider/file_watch.go]: apologies. i somehow removed it during cleanup. added it back.

File: config/configmapprovider/file.go [hunk 7]
```diff
@@ -18,12 +18,17 @@ import (
...
+	"time"
 
 	"go.opentelemetry.io/collector/config"
 )
 
+var errNoOp = errors.New("no change")
+
 type fileMapProvider struct {
 	fileName string
+	watching bool
```
Comment (tigrannajaryan) [inline config/configmapprovider/file.go:31]: What's the purpose of this? It does't seem to be set anywhere.

File: service/config_watcher.go [hunk 8]
```diff
@@ -62,7 +68,6 @@ func (cm *configWatcher) onChange(event *configmapprovider.ChangeEvent) {
 	}
 }
 
-func (cm *configWatcher) close(ctx context.Context) error {
+func (cm *configWatcher) close() {
 	close(cm.watcher)
-	return cm.ret.Close(ctx)
```
Comment (tigrannajaryan) [inline service/config_watcher.go:67]: Why is this deleted?

**Comment by bogdandrutu** [comment]
Please rebase, and mark as resolved comments that are resolved.

File: config/configmapprovider/file.go (hunk 7 above)
Comment (vjsamuel) [reply config/configmapprovider/file.go:31]: fixed this one. this is to ensure that we create a watch only once as `Retrieve` is called each time the `onChange` is invoked during a file change.

File: service/config_watcher.go (hunk 8 above)
Comment (vjsamuel) [reply service/config_watcher.go:67]: i have modified the code flow in a way that the config watcher is only created once in the lifecycle of the collector as compared to how it was originally implemented where a config watcher is created per change in the config file. once that change was made, it didnt make sense to close the retrieved and pass the error down. i moved that logic into a `get()` method that follows the `Retrieve()` -> `watch` -> `Close()` lifecycle.

File: config/configmapprovider/file.go [hunk 9]
```diff
@@ -43,9 +48,65 @@ func (fmp *fileMapProvider) Retrieve(_ context.Context, _ func(*ChangeEvent)) (R
 		return nil, fmt.Errorf("error loading config file %q: %w", fmp.fileName, err)
 	}
 
+	// Ensure that watches are only created once as Retrieve is called each time that `onChange()` is invoked.
+	if !fmp.watching {
+		watchFile(ctx, fmp.fileName, onChange)
+		fmp.watching = true
+	}
+
 	return &simpleRetrieved{confMap: cp}, nil
```
Comment (tigrannajaryan) [inline config/configmapprovider/file.go:57]: I don't this will as expected by the `Provider` interface. The `Retrieved` that is returned is expected to implement a `Close` function that stops watching and guarantees that `onChange` will not be called after that. See https://github.com/open-telemetry/opentelemetry-collector/blob/96a882a87742e2a921ebabf8e43ad97d8bf0e8c0/config/configmapprovider/provider.go#L76

File: service/config_watcher.go (hunk 8 above)
Comment (tigrannajaryan) [reply service/config_watcher.go:67]: >i have modified the code flow in a way that the config watcher is only created once in the lifecycle of the collector

The new rearranged code is harder to follow and understand. Please refactor it to clearly show that the code follows lifecycle described in the `Provider` [comments](https://github.com/open-telemetry/opentelemetry-collector/blob/96a882a87742e2a921ebabf8e43ad97d8bf0e8c0/config/configmapprovider/provider.go#L55):
```
// The typical usage is the following:
//
//		r := mapProvider.Retrieve()
//		r.Get()
//		// wait for onChange() to be called.
//		r.Close()
//		r = mapProvider.Retrieve()
//		r.Get()
//		// wait for onChange() to be called.
//		r.Close()
//		// repeat Retrieve/Get/wait/Close cycle until it is time to shut down the Collector process.
//		// ...
//		mapProvider.Shutdown()
```
It was more visible before this change, admittedly it was not ideal but was better than what we have now. Now it is even harder to see that we are actually following the required lifecycle. All the current loop in runAndWaitForShutdownEvent shows is a watch, followed by `get()`.

**Comment by seh** [comment]
> This PR doesn't use an FS notify/inotify style watcher as we have seen that Kubernetes doesnt support such watches when a config is mounted as a config file.

Can you clarify why this doesn't work? Tools like [_jimmidyson/configmap-reload_](https://github.com/jimmidyson/configmap-reload) attempt to detect such changes, as does [the Thanos reloader](https://github.com/thanos-io/thanos/tree/main/pkg/reloader) used with Prometheus.

**Comment by seh** [comment]
This capability could also help with #1591.

**Review by HankVeal12 — COMMENTED** [review]
[https://www.wellsfargo.com/preselected/?lang=en](Google)
